### ISSUE TITLE
Accessibility Support

### ISSUE THREAD

**Issue by cary-rowen** [dail8859/NotepadNext#309]
Hi,

I am a screen reader user and the current version of NotepadNext does not work with the NVDA screen reader.
NVDA is a completely open source and free screen reader for the windows platform.
You can learn about NVDA [here](https://github.com/nvaccess/nvda)

Many editors using Scintilla NVDA support is good, such as Notepad++.

Can you look into this?

Thanks

**Comment by inkydragon** [comment]
I can reproduce the issue with NVDA 2022.3 + npn V0.6

The menu bar is accessible via the ALT key.
But the main editor panel cannot be recognized by screen readers.
A simple test is to select a word in the editor and see if the screen reader outputs it properly.

**Comment by dail8859** [comment]
I know nothing when it comes to accessibility support, so I can't even begin to guess what needs done.

I know Scintilla on the Win32 framework supports it (i.e. Notepad++) but since this is built on Qt I'm not sure if this is a Scintilla issue, or something that just needs enabled in Qt. I'd definitely love to see this kind of functionality in the application. If I get some free time I'll try to look into it a bit.

**Comment by inkydragon** [comment]
[ScintillaDoc#Accessibility](https://www.scintilla.org/ScintillaDoc.html#Accessibility)

It looks like the only accessibility related function in Scintilla is `[get|set]accessibility`.
I'm not sure where to call it, though.

A failed attempt:
```diff
diff --git a/src/NotepadNext/ScintillaNext.cpp b/src/NotepadNext/ScintillaNext.cpp
index 0ee37e3..2879ee4 100644
--- a/src/NotepadNext/ScintillaNext.cpp
+++ b/src/NotepadNext/ScintillaNext.cpp
@@ -59,6 +59,11 @@ ScintillaNext::ScintillaNext(QString name, QWidget *parent) :
     indicatorResources.disableRange(0, 7);
     indicatorResources.disableRange(INDICATOR_IME, INDICATOR_IME_MAX);
     indicatorResources.disableRange(INDICATOR_HISTORY_REVERTED_TO_ORIGIN_INSERTION, INDICATOR_HISTORY_REVERTED_TO_MODIFIED_DELETION);
+
+    // set a11y
+    qInfo("Old Accessibility: %d", (accessibility()));
+    setAccessibility((Scintilla::sptr_t) Accessibility::Enabled);
+    qInfo("New Accessibility: %d", (accessibility()));
 }
```

**Comment by dail8859** [comment]
Looks like Scintilla's default implementation leaves it up to the platform implementation: 

https://github.com/dail8859/NotepadNext/blob/73a7591fa25bf945832e1eef1496337b75442100/src/scintilla/src/Editor.cxx#L8093-L8095

Looking at the Qt layer it doesn't handle it at all in here:

https://github.com/dail8859/NotepadNext/blob/73a7591fa25bf945832e1eef1496337b75442100/src/scintilla/qt/ScintillaEditBase/ScintillaQt.cpp#L736-L774

Scintilla documentation also doesn't mention it being supported in Qt.

That's at least an answer why it doesn't work currently.

**Comment by inkydragon** [comment]
> Looking at the Qt layer it doesn't handle it at all in here:.

So, this becomes an `upstream` issue.
To make NpN accessible, we need to fix Scintilla's Qt backend first.

**Comment by dail8859** [comment]
> So, this becomes an upstream issue.
To make NpN accessible, we need to fix Scintilla's Qt backend first.

Yes that seems to be the case.

> Perhaps only the Gtk backend is accessible currently.

I think cocoa may as well. NOtepad++ does it so there seems to be some sort of support on Win32 but not sure if that is via Scintilla itself or some other means.

**Comment by dail8859** [comment]
Closing this issue for now since there's nothing directly Notepad Next can do. Accessibility support is definitely desirable but something the Scintilla project would have to incorporate, or if anyone else had some other ideas. You could discuss it on their mailing list or bug tracker but I highly doubt anyone will start actively working on it.

**Comment by inkydragon** [comment]
> I highly doubt anyone will start actively working on it.


> Neil Hodgson- 2016-02-18:  
> Accessibility would be a good addition to Scintilla. However, it is a large amount of work made harder by each platform having a different API for accessibility. It is unlikely that I would be able to work on this without customers providing funding.
> 
> https://sourceforge.net/p/scintilla/feature-requests/1137/

**Comment by cary-rowen** [comment]
This sounds like a sad thing.
